# Module 1: Data Loading  and exploration

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("Titanic-Dataset.csv")
df

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


## 1. Parameter : nrows

Load only first N rows.

In [5]:
pd.read_csv("Titanic-Dataset.csv",nrows=100)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
95,96,0,3,"Shorney, Mr. Charles Joseph",male,NaN,0,0,374910,8.0500,NaN,S
96,97,0,1,"Goldschmidt, Mr. George B",male,71.0,0,0,PC 17754,34.6542,A5,C
97,98,1,1,"Greenfield, Mr. William Bertram",male,23.0,0,1,PC 17759,63.3583,D10 D12,C
98,99,1,2,"Doling, Mrs. John T (Ada Julia Bone)",female,34.0,0,1,231919,23.0000,NaN,S


## 2. usecols

Load selected columns

In [8]:
pd.read_csv(
    "Titanic-Dataset.csv",
    usecols=["Survived","Pclass"]
)

,Survived,Pclass
0,0,3
1,1,1
2,1,3
3,1,1
4,0,3
...,...,...
886,0,2
887,1,1
888,0,3
889,1,1


## 3. index_cols

Set columns as index

In [9]:
pd.read_csv(
    "Titanic-Dataset.csv",
    index_col=["Pclass"]
)

,PassengerId,Survived,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
Pclass,,,,,,,,,,,
3,1,0,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,3,1,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
1,4,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
3,5,0,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...
2,887,0,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
1,888,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
3,889,0,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S


##  4. skiprows 

Skip rows.

In [17]:
pd.read_csv(
    "Titanic-Dataset.csv",
    skiprows=1
)

,1,0,3,"Braund, Mr. Owen Harris",male,22,1.1,0.1,A/5 21171,7.25,Unnamed: 10,S
0,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
1,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
2,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
3,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
4,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
...,...,...,...,...,...,...,...,...,...,...,...,...
885,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
886,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
887,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
888,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


## 5. navalues

Custom missing values.

Convert all into NaN.


```
df = pd.read_csv(
    "file.csv",
    na_values=["NA","Missing","Null"]
)
```

In [19]:
pd.read_csv(
    "Titanic-Dataset.csv",
    na_values=["NaN","missing"]
)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


## 6. parse_dates

Convert date columns automatically.

Without:

OrderDate
2025-01-01

Loaded as:

object
```
With:

df = pd.read_csv(
    "sales.csv",
    parse_dates=["OrderDate"]
)
```
Loaded as:

datetime64

In [20]:
type(df)

pandas.core.frame.DataFrame

In [21]:
df.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


## Topic 2: First Look at Data

### 1.Head()

Shows first rows of the DataFrame.

In [22]:
df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


### 2. tail()

Shows last rows of dataset.

In [23]:
df.tail(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


### 4. sample()

Purpose

Returns random rows.

In [29]:
df.sample()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
447,448,1,1,"Seward, Mr. Frederic Kimber",male,34.0,0,0,113794,26.55,NaN,S


In [30]:
df.sample(4)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
740,741,1,1,"Hawksford, Mr. Walter James",male,NaN,0,0,16988,30.000,D45,S
684,685,0,2,"Brown, Mr. Thomas William Solomon",male,60.0,1,1,29750,39.000,NaN,S
425,426,0,3,"Wiseman, Mr. Phillippe",male,NaN,0,0,A/4. 34244,7.250,NaN,S
761,762,0,3,"Nirva, Mr. Iisakki Antino Aijo",male,41.0,0,0,SOTON/O2 3101272,7.125,NaN,S


### Now output remains same.
```
df.sample(
    n=5,
    random_state=42
)
```

Why Industry Uses It

When:

1. debugging
2. reporting
3. sharing notebooks
4. reproducing results

everyone should see same sample.

In [33]:
df.sample(
    n=5,
    random_state=42
)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
709,710,1,3,"Moubarek, Master. Halim Gonios (""William George"")",male,NaN,1,1,2661,15.2458,NaN,C
439,440,0,2,"Kvillner, Mr. Johan Henrik Johannesson",male,31.0,0,0,C.A. 18723,10.5000,NaN,S
840,841,0,3,"Alhomaki, Mr. Ilmari Rudolf",male,20.0,0,0,SOTON/O2 3101287,7.9250,NaN,S
720,721,1,2,"Harper, Miss. Annie Jessie ""Nina""",female,6.0,0,1,248727,33.0000,NaN,S
39,40,1,3,"Nicola-Yarred, Miss. Jamila",female,14.0,1,0,2651,11.2417,NaN,C


```
Parameter: frac

Percentage of rows.
```

In [38]:
df.sample(
    
    frac=0.25
)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
74,75,1,3,"Bing, Mr. Lee",male,32.0,0,0,1601,56.4958,NaN,S
149,150,0,2,"Byles, Rev. Thomas Roussel Davids",male,42.0,0,0,244310,13.0000,NaN,S
380,381,1,1,"Bidois, Miss. Rosalie",female,42.0,0,0,PC 17757,227.5250,NaN,C
734,735,0,2,"Troupiansky, Mr. Moses Aaron",male,23.0,0,0,233639,13.0000,NaN,S
425,426,0,3,"Wiseman, Mr. Phillippe",male,NaN,0,0,A/4. 34244,7.2500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
878,879,0,3,"Laleff, Mr. Kristo",male,NaN,0,0,349217,7.8958,NaN,S
840,841,0,3,"Alhomaki, Mr. Ilmari Rudolf",male,20.0,0,0,SOTON/O2 3101287,7.9250,NaN,S
690,691,1,1,"Dick, Mr. Albert Adrian",male,31.0,1,0,17474,57.0000,B20,S
549,550,1,2,"Davies, Master. John Morgan Jr",male,8.0,1,1,C.A. 33112,36.7500,NaN,S


## Topic 3: Understanding data structure

### 1. shape

Return Type

Returns a tuple:




Before analysis, you need to know dataset size.

Questions like:

1. How many rows?
2. How many columns?
3. Is dataset small or huge?

are answered instantly by shape.

In [1]:
import pandas as pd 

In [4]:
df = pd.read_csv("Titanic-Dataset.csv")

In [3]:
df.shape

(891, 12)

### Accessing Individually


#### Rows

In [4]:
df.shape[0]

891

#### Columns

In [7]:
df.shape[1]

12

In [8]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 891
Columns: 12


### 2. Columns

In [9]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

#### 2.1 convert to list

In [10]:
list(df.columns)

['PassengerId',
 'Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Cabin',
 'Embarked']

#### 2.2 Count columns

In [2]:
import pandas as pd 

In [6]:
len(df.columns)

12

### 3. index

Every row needs an identifier.
```
Output
RangeIndex(
start=0,
stop=891,
step=1
)

Meaning:

Starts at 0
Ends before 891
Step size 1
```

In [7]:
df.index

RangeIndex(start=0, stop=891, step=1)

### len(index)

In [8]:
len(df.index)

891

## Topic 4: info()

1. Why Does info() Exist?
- - What kind of data do I have?"

Suppose you load:

df = pd.read_csv("employees.csv")

Questions:

1. How many rows are there?
2. How many columns?
3. Which columns have missing values?
4. Which columns are numeric?
5. Which columns are text?
6. How much memory is being used?

info() answers all of these at once.

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


## Topic 5: describe()

"What do the values look like?"

Gives Statistical summary.

```
Default
df.describe()

Only numeric columns.

Include Everything
df.describe(include='all')
```

In [11]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [13]:
df.describe(include="all")

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
count,891.000000,891.000000,891.000000,891,891,714.000000,891.000000,891.000000,891,891.000000,204,889
unique,NaN,NaN,NaN,891,2,NaN,NaN,NaN,681,NaN,147,3
top,NaN,NaN,NaN,"Dooley, Mr. Patrick",male,NaN,NaN,NaN,347082,NaN,G6,S
freq,NaN,NaN,NaN,1,577,NaN,NaN,NaN,7,NaN,4,644
mean,446.000000,0.383838,2.308642,NaN,NaN,29.699118,0.523008,0.381594,NaN,32.204208,NaN,NaN
std,257.353842,0.486592,0.836071,NaN,NaN,14.526497,1.102743,0.806057,NaN,49.693429,NaN,NaN
min,1.000000,0.000000,1.000000,NaN,NaN,0.420000,0.000000,0.000000,NaN,0.000000,NaN,NaN
25%,223.500000,0.000000,2.000000,NaN,NaN,20.125000,0.000000,0.000000,NaN,7.910400,NaN,NaN
50%,446.000000,0.000000,3.000000,NaN,NaN,28.000000,0.000000,0.000000,NaN,14.454200,NaN,NaN
75%,668.500000,1.000000,3.000000,NaN,NaN,38.000000,1.000000,0.000000,NaN,31.000000,NaN,NaN


### 5.1 count



count != total rows

This immediately reveals missing values.

### 5.2 std (Standard Deviation)

How spread out are values?

1. Dataset A:
```
[29,30,31]

Very close together.

Small std.
```

2. Dataset B:
```
[5,30,90]

Very spread out.

Large std.
```

Interpretation:

- Small std → values clustered
- Large std → values scattered

### 5.3 min/max

Smallest value. and largest value 

### 5.4. Quartiles

Imagine sorted ages:
```
10 15 18 20 22 25 28 30 35 40 50
25% ---------- 50% --------- 75%
```
These values help understand distribution.



#### a) 25%
25% of observations are below this value.

```
Example:

Age 25% = 20

Meaning:

25% passengers are younger than 20.
```

#### b) 50%

Also called:

Median

```
Example:

Age 50% = 28

Meaning:

Half of passengers are younger than 28.
Half are older than 28.
```

#### c) 75%

Example:

Age 75% = 38

Meaning:

75% passengers are younger than 38. 

# Module 2: Selecting Data

## Topic 1: Single Column Selection

Single Column Selection
```
Syntax:

df["Age"]
```
Returns-->  Series
```
Recommended:

df["Age"]
```
Avoid-->  df.Age

In [1]:
import pandas as pd 

In [2]:
df = pd.read_csv("Titanic-Dataset.csv")

In [3]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [6]:
age = df["Age"]
age

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

1. type of age

In [7]:
type(age)

pandas.core.series.Series

2. mean

In [9]:
age.mean()

np.float64(29.69911764705882)

3. max

In [10]:
age.max()

np.float64(80.0)

4. min

In [11]:
age.min()

np.float64(0.42)

5. unique

In [ ]:
age.unique()

array([22.  , 38.  , 26.  , 35.  ,   nan, 54.  ,  2.  , 27.  , 14.  ,
        4.  , 58.  , 20.  , 39.  , 55.  , 31.  , 34.  , 15.  , 28.  ,
        8.  , 19.  , 40.  , 66.  , 42.  , 21.  , 18.  ,  3.  ,  7.  ,
       49.  , 29.  , 65.  , 28.5 ,  5.  , 11.  , 45.  , 17.  , 32.  ,
       16.  , 25.  ,  0.83, 30.  , 33.  , 23.  , 24.  , 46.  , 59.  ,
       71.  , 37.  , 47.  , 14.5 , 70.5 , 32.5 , 12.  ,  9.  , 36.5 ,
       51.  , 55.5 , 40.5 , 44.  ,  1.  , 61.  , 56.  , 50.  , 36.  ,
       45.5 , 20.5 , 62.  , 41.  , 52.  , 63.  , 23.5 ,  0.92, 43.  ,
       60.  , 10.  , 64.  , 13.  , 48.  ,  0.75, 53.  , 57.  , 80.  ,
       70.  , 24.5 ,  6.  ,  0.67, 30.5 ,  0.42, 34.5 , 74.  ])

6. missing values

In [17]:
age.isnull().sum()

np.int64(177)

## Topic 2: Multiple Column Selection

Return Type

DataFrame

In [20]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [22]:
type(df[["Age"]])

pandas.core.frame.DataFrame

In [23]:
df[["Sex","Age","Parch"]]

,Sex,Age,Parch
0,male,22.0,0
1,female,38.0,0
2,female,26.0,0
3,female,35.0,0
4,male,35.0,0
...,...,...,...
886,male,27.0,0
887,female,19.0,0
888,female,NaN,2
889,male,26.0,0


In [25]:
df[["Age","Age"]]

,Age,Age
0,22.0,22.0
1,38.0,38.0
2,26.0,26.0
3,35.0,35.0
4,35.0,35.0
...,...,...
886,27.0,27.0
887,19.0,19.0
888,NaN,NaN
889,26.0,26.0


In [26]:
df[["Age"]].shape

(891, 1)

## Topic 3: Understanding Rows and Indexes

In [4]:
df = pd.read_csv("Titanic-Dataset.csv")

In [27]:
df.head(4)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S


In [28]:
df.index

RangeIndex(start=0, stop=891, step=1)

In [2]:
import pandas as pd

### 1. loc

#### 1.1 .loc[row]

In [10]:
(df.loc[4])

PassengerId                           5
Survived                              0
Pclass                                3
Name           Allen, Mr. William Henry
Sex                                male
Age                                35.0
SibSp                                 0
Parch                                 0
Ticket                           373450
Fare                               8.05
Cabin                               NaN
Embarked                              S
Name: 4, dtype: object

#### 1.2  .loc[[multiple rows]]

In [26]:
df.loc[[1,2,3]]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S


#### 1.3  loc[row,column]

In [13]:
df.loc[0, "PassengerId"]

np.int64(1)

#### 1.4 .loc[[rows],[column]]

In [15]:
df.loc[[0,1],["PassengerId","Survived"]]

,PassengerId,Survived
0,1,0
1,2,1


### 2. .iloc

In [20]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


#### 2.1 .iloc[rows]

In [18]:
print(df.iloc[0])

PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                               22.0
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object


#### 2.2 .iloc[[multiple rows]]

In [28]:
df.iloc[[0,4,5]]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q


#### 2.3 .iloc[rows,column]

In [23]:
(df.iloc[0,3])

'Braund, Mr. Owen Harris'

#### 2.4 .iloc[[rows],[column]]

In [25]:
df.iloc[[0,1,4],[0,6]]

,PassengerId,SibSp
0,1,1
1,2,1
4,5,0


### 3. slicing 

#### 3.1 using loc

In [29]:
df.head(4)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S


##### a) row slicing 

In [31]:
df.loc[0:5]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q


##### b) row + column slicing 

In [34]:
df.loc[0:5,["PassengerId","Sex"]]

,PassengerId,Sex
0,1,male
1,2,female
2,3,female
3,4,female
4,5,male
5,6,male


In [36]:
df.loc[0:5,"Survived":"Age"]

,Survived,Pclass,Name,Sex,Age
0,0,3,"Braund, Mr. Owen Harris",male,22.0
1,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0
2,1,3,"Heikkinen, Miss. Laina",female,26.0
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0
4,0,3,"Allen, Mr. William Henry",male,35.0
5,0,3,"Moran, Mr. James",male,NaN


##### c) fancy indexing Slicing

In [37]:
df.loc[[5,3,4],"Survived":"Sex"]

,Survived,Pclass,Name,Sex
5,0,3,"Moran, Mr. James",male
3,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female
4,0,3,"Allen, Mr. William Henry",male


#### 3.2 using iloc

In [41]:
df.head(4)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S


##### a) row slicing

In [39]:
df.iloc[3:7]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S


##### b) row + column slicing 

In [40]:
df.iloc[0:3,3:8]

,Name,Sex,Age,SibSp,Parch
0,"Braund, Mr. Owen Harris",male,22.0,1,0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0
2,"Heikkinen, Miss. Laina",female,26.0,0,0


##### c) fancy slicing 

In [42]:
df.iloc[[3,4,5],[8,9,10]]

,Ticket,Fare,Cabin
3,113803,53.1000,C123
4,373450,8.0500,NaN
5,330877,8.4583,NaN


# Module 3: Filtering Data

Filtering means:-

Examples:

1. Show passengers older than 30
2. Show females only
3. Show tickets with Fare > 100
4. Show passengers who survived
5. Show passengers older than 30 AND female

This is filtering.

## Topic 1: Single Condition Filtering


```
df.loc[
    condition,
    columns
]
```


In [43]:
df.head(4)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S


In [ ]:
df["Age"] >= 22  # The result will be total of 891 rows 

0       True
1       True
2       True
3       True
4       True
       ...  
886     True
887    False
888    False
889     True
890     True
Name: Age, Length: 891, dtype: bool

In [56]:
print(df[df["Age"] >= 22].shape[0] )
(df[df["Age"] >= 22] ) # Now, the result will be only matching to condition.  

510


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
884,885,0,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [46]:
df["Sex"] == "male"

0       True
1      False
2      False
3      False
4       True
       ...  
886     True
887    False
888    False
889     True
890     True
Name: Sex, Length: 891, dtype: bool

In [60]:
print(df[df["Sex"] == "male"].shape[0])
df[df["Sex"] == "male"]

577


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
883,884,0,2,"Banfield, Mr. Frederick James",male,28.0,0,0,C.A./SOTON 34068,10.5000,NaN,S
884,885,0,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [81]:
df[(df["Survived"] == 1)][['Name',"Survived"]]

,Name,Survived
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1
2,"Heikkinen, Miss. Laina",1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1
8,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",1
9,"Nasser, Mrs. Nicholas (Adele Achem)",1
...,...,...
875,"Najib, Miss. Adele Kiamie ""Jane""",1
879,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",1
880,"Shelley, Mrs. William (Imanita Parrish Hall)",1
887,"Graham, Miss. Margaret Edith",1


In [74]:
df[df["Age"] == 25][["Name","Age"]]

,Name,Age
75,"Moen, Mr. Sigurd Hansen",25.0
134,"Sobey, Mr. Samuel James Hayden",25.0
246,"Lindahl, Miss. Agda Thorilda Viktoria",25.0
267,"Persson, Mr. Ernst Ulrik",25.0
271,"Tornquist, Mr. William Henry",25.0
343,"Sedgwick, Mr. Charles Frederick Waddington",25.0
353,"Arnold-Franchi, Mr. Josef",25.0
370,"Harder, Mr. George Achilles",25.0
442,"Petterson, Mr. Johan Emil",25.0
484,"Bishop, Mr. Dickinson H",25.0


```
Show only Name and Fare of passengers whose Age > 60.
```

In [95]:
df[df["Age"] > 60][["Name","Fare","Age"]]

,Name,Fare,Age
33,"Wheadon, Mr. Edward H",10.5000,66.0
54,"Ostby, Mr. Engelhart Cornelius",61.9792,65.0
96,"Goldschmidt, Mr. George B",34.6542,71.0
116,"Connors, Mr. Patrick",7.7500,70.5
170,"Van der hoef, Mr. Wyckoff",33.5000,61.0
252,"Stead, Mr. William Thomas",26.5500,62.0
275,"Andrews, Miss. Kornelia Theodosia",77.9583,63.0
280,"Duane, Mr. Frank",7.7500,65.0
326,"Nysveen, Mr. Johan Hansen",6.2375,61.0
438,"Fortune, Mr. Mark",263.0000,64.0


In [112]:
df.loc[
    df["Age"]>60,
    ["Name","Fare"]
]

,Name,Fare
33,"Wheadon, Mr. Edward H",10.5000
54,"Ostby, Mr. Engelhart Cornelius",61.9792
96,"Goldschmidt, Mr. George B",34.6542
116,"Connors, Mr. Patrick",7.7500
170,"Van der hoef, Mr. Wyckoff",33.5000
252,"Stead, Mr. William Thomas",26.5500
275,"Andrews, Miss. Kornelia Theodosia",77.9583
280,"Duane, Mr. Frank",7.7500
326,"Nysveen, Mr. Johan Hansen",6.2375
438,"Fortune, Mr. Mark",263.0000


```
Easy Rule
One Row + One Column
df.loc[0, "Age"]

➡️ Scalar

Multiple Rows + One Column
df.loc[0:5, "Age"]

➡️ Series

Multiple Rows + Multiple Columns
df.loc[0:5, ["Age", "Fare"]]

➡️ DataFrame

One Row + Multiple Columns
df.loc[0, ["Age", "Fare"]]

➡️ Series
```

## Topic 2: Multiple Condition Filtering

In [1]:
import pandas as  pd 

In [2]:
df = pd.read_csv("Titanic-Dataset.csv")

In [11]:
df.loc[(df["Age"] > 30)  & (df["Fare"] > 100)]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
195,196,1,1,"Lurette, Miss. Elise",female,58.0,0,0,PC 17569,146.5208,B80,C
215,216,1,1,"Newell, Miss. Madeleine",female,31.0,1,0,35273,113.2750,D36,C
258,259,1,1,"Ward, Miss. Anna",female,35.0,0,0,PC 17755,512.3292,NaN,C
268,269,1,1,"Graham, Mrs. William Thompson (Edith Junkins)",female,58.0,0,1,PC 17582,153.4625,C125,S
269,270,1,1,"Bissette, Miss. Amelia",female,35.0,0,0,PC 17760,135.6333,C99,S
299,300,1,1,"Baxter, Mrs. James (Helene DeLaudeniere Chaput)",female,50.0,0,1,PC 17558,247.5208,B58 B60,C
318,319,1,1,"Wick, Miss. Mary Natalie",female,31.0,0,2,36928,164.8667,C7,S
319,320,1,1,"Spedden, Mrs. Frederic Oakley (Margaretta Corn...",female,40.0,1,1,16966,134.5000,E34,C
325,326,1,1,"Young, Miss. Marie Grice",female,36.0,0,0,PC 17760,135.6333,C32,C
332,333,0,1,"Graham, Mr. George Edward",male,38.0,0,1,PC 17582,153.4625,C91,S


```
Passengers who are female AND survived.
Return only:

Name
Age
Fare
```

In [15]:
df.loc[
    (df['Sex'] == "female") &
    (df["Survived"] == 1),
    ["Name","Age","Fare"]
    ]
# df[(df['Sex'] == "female") & (df["Survived"] == 1)].shape[0]

,Name,Age,Fare
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,71.2833
2,"Heikkinen, Miss. Laina",26.0,7.9250
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,53.1000
8,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",27.0,11.1333
9,"Nasser, Mrs. Nicholas (Adele Achem)",14.0,30.0708
...,...,...,...
874,"Abelson, Mrs. Samuel (Hannah Wizosky)",28.0,24.0000
875,"Najib, Miss. Adele Kiamie ""Jane""",15.0,7.2250
879,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",56.0,83.1583
880,"Shelley, Mrs. William (Imanita Parrish Hall)",25.0,26.0000


```
Show passengers whose:

Age > 60
OR
Fare > 100
```

In [16]:
df[
    (df["Age"] > 60)
    | (df["Fare"] >100)

    
]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
27,28,0,1,"Fortune, Mr. Charles Alexander",male,19.0,3,2,19950,263.0000,C23 C25 C27,S
31,32,1,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,NaN,1,0,PC 17569,146.5208,B78,C
33,34,0,2,"Wheadon, Mr. Edward H",male,66.0,0,0,C.A. 24579,10.5000,NaN,S
54,55,0,1,"Ostby, Mr. Engelhart Cornelius",male,65.0,0,1,113509,61.9792,B30,C
88,89,1,1,"Fortune, Miss. Mabel Helen",female,23.0,3,2,19950,263.0000,C23 C25 C27,S
...,...,...,...,...,...,...,...,...,...,...,...,...
779,780,1,1,"Robert, Mrs. Edward Scott (Elisabeth Walton Mc...",female,43.0,0,1,24160,211.3375,B3,S
802,803,1,1,"Carter, Master. William Thornton II",male,11.0,1,2,113760,120.0000,B96 B98,S
829,830,1,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0000,B28,NaN
851,852,0,3,"Svensson, Mr. Johan",male,74.0,0,0,347060,7.7750,NaN,S


```
Q1

Passengers who are:

Male
OR
Fare > 100

Return:

Name
Sex
Fare

using loc.
```

In [17]:
df.loc[
    (df["Sex"] == "Male")
    | (df["Fare"] > 100),
    ["Name","Sex","Fare"]
]

,Name,Sex,Fare
27,"Fortune, Mr. Charles Alexander",male,263.0000
31,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,146.5208
88,"Fortune, Miss. Mabel Helen",female,263.0000
118,"Baxter, Mr. Quigg Edmond",male,247.5208
195,"Lurette, Miss. Elise",female,146.5208
215,"Newell, Miss. Madeleine",female,113.2750
258,"Ward, Miss. Anna",female,512.3292
268,"Graham, Mrs. William Thompson (Edith Junkins)",female,153.4625
269,"Bissette, Miss. Amelia",female,135.6333
297,"Allison, Miss. Helen Loraine",female,151.5500


## Topic 3: isin()

1. Check whether values belong to a list.

```
Syntax
df["column"].isin([...])
```

Suppose you want: Passengers whose Pclass is: 1 OR 2

```
You could write:

df[
    (df["Pclass"] == 1)
    |
    (df["Pclass"] == 2)
]
```
Works.
But now imagine:

Pclass = 1 OR 2 OR 3 OR 4 OR 5

Now your code starts looking like a conspiracy wall.

Solution: isin()
```
df["Pclass"].isin([1, 2])

Output:

True
True
False
True
...
```
A Boolean Series.

In [ ]:
"Example 1"

df[df["Pclass"].isin([1,2])]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
...,...,...,...,...,...,...,...,...,...,...,...,...
880,881,1,2,"Shelley, Mrs. William (Imanita Parrish Hall)",female,25.0,0,1,230433,26.0000,NaN,S
883,884,0,2,"Banfield, Mr. Frederick James",male,28.0,0,0,C.A./SOTON 34068,10.5000,NaN,S
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S


In [25]:
"Example 2"


df[df["Embarked"].isin(["C"])]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C
19,20,1,3,"Masselmani, Mrs. Fatima",female,NaN,0,0,2649,7.2250,NaN,C
26,27,0,3,"Emir, Mr. Farred Chehab",male,NaN,0,0,2631,7.2250,NaN,C
30,31,0,1,"Uruchurtu, Don. Manuel E",male,40.0,0,0,PC 17601,27.7208,NaN,C
...,...,...,...,...,...,...,...,...,...,...,...,...
866,867,1,2,"Duran y More, Miss. Asuncion",female,27.0,1,0,SC/PARIS 2149,13.8583,NaN,C
874,875,1,2,"Abelson, Mrs. Samuel (Hannah Wizosky)",female,28.0,1,0,P/PP 3381,24.0000,NaN,C
875,876,1,3,"Najib, Miss. Adele Kiamie ""Jane""",female,15.0,0,0,2667,7.2250,NaN,C
879,880,1,1,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",female,56.0,0,1,11767,83.1583,C50,C


### using with loc

In [32]:
"Example 3"

df.loc[
    df["Embarked"].isin(["C"]),
    ["Name"]

]

,Name
1,"Cumings, Mrs. John Bradley (Florence Briggs Th..."
9,"Nasser, Mrs. Nicholas (Adele Achem)"
19,"Masselmani, Mrs. Fatima"
26,"Emir, Mr. Farred Chehab"
30,"Uruchurtu, Don. Manuel E"
...,...
866,"Duran y More, Miss. Asuncion"
874,"Abelson, Mrs. Samuel (Hannah Wizosky)"
875,"Najib, Miss. Adele Kiamie ""Jane"""
879,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)"


### Not in

In [33]:
df[
    ~(df["Pclass"].isin([3]))
]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
...,...,...,...,...,...,...,...,...,...,...,...,...
880,881,1,2,"Shelley, Mrs. William (Imanita Parrish Hall)",female,25.0,0,1,230433,26.0000,NaN,S
883,884,0,2,"Banfield, Mr. Frederick James",male,28.0,0,0,C.A./SOTON 34068,10.5000,NaN,S
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S


## Topic 4: between()

Suppose you want:  Passengers whose Age is between 20 and 40
```
between(start, end)
By default:

Both boundaries included.

Meaning:

20 <= Age <= 40
```

In [35]:
"Example 1"

df[df["Age"].between(20,50)]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
884,885,0,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


### Combining With Other Conditions

- &
- |
- loc
- iloc
etc

In [43]:
"Example 1"

df.loc[
    (df["Fare"] > 200)
    & 
    (df["Age"].between(1,18))
]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
311,312,1,1,"Ryerson, Miss. Emily Borie",female,18.0,2,2,PC 17608,262.3750,B57 B59 B63 B66,C
689,690,1,1,"Madill, Miss. Georgette Alexandra",female,15.0,0,1,24160,211.3375,B5,S
700,701,1,1,"Astor, Mrs. John Jacob (Madeleine Talmadge Force)",female,18.0,1,0,PC 17757,227.5250,C62 C64,C


In [ ]:
df.loc[

    ((df["Sex"] == "female")

      &

      (df["Age"] > 30) ),
      



]

,Name,Age
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0
11,"Bonnell, Miss. Elizabeth",58.0
15,"Hewlett, Mrs. (Mary D Kingcome)",55.0
18,"Vander Planke, Mrs. Julius (Emelia Maria Vande...",31.0
...,...,...
862,"Swift, Mrs. Frederick Joel (Margaret Welles Ba...",48.0
865,"Bystrom, Mrs. (Karolina)",42.0
871,"Beckwith, Mrs. Richard Leonard (Sallie Monypeny)",47.0
879,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",56.0


# Topic 5: value_counts()

- Count frequency of unique values.


Many columns contain categories.

Examples:

- Sex
- Pclass
- Embarked
- Department
- City
- Country
- Product Category

Analysts constantly want to know:

How many times does each category occur?

That's exactly what value_counts() answers.

```
Syntax
df["column"].value_counts()
```

```
Return Type -> pandas.Series
```



df["Sex"].value_counts()

will return-

- male      577
- female    314

Because Pandas sorts by count descending by default.

In [2]:
import pandas as pd

In [8]:
df = pd.read_csv("Titanic-Dataset.csv")

df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


In [7]:
"Example 1"

df["Pclass"].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [13]:
"Example 2"

df["Sex"].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

## 1. nomalise gives proportion

In [15]:
df["Sex"].value_counts(normalize=True) *100

Sex
male      64.758698
female    35.241302
Name: proportion, dtype: float64

## 2. dropna() = False

In [24]:
df["Cabin"].value_counts(dropna=True)

Cabin
G6             4
C23 C25 C27    4
B96 B98        4
F2             3
D              3
              ..
E17            1
A24            1
C50            1
B42            1
C148           1
Name: count, Length: 147, dtype: int64

In [25]:
df["Cabin"].value_counts(dropna=False)

Cabin
NaN            687
G6               4
C23 C25 C27      4
B96 B98          4
F2               3
              ... 
E17              1
A24              1
C50              1
B42              1
C148             1
Name: count, Length: 148, dtype: int64